### ISSUE TITLE
Why is the implementation of Mamba so slow?

### ISSUE THREAD

**Issue by baisonmg** [johnma2006/mamba-minimal#21]
I haven't run the official version of Mamba, but I've run your implementation, and it seems that the training speed of this model is much slower than that of the Transformer.

**Comment by aryanmangal769** [comment]
I am also facing the similar issue. Are you able to find the reason for it?

**Comment by lingxitong** [comment]
that is mainly because there is something wrong which lead the loss.backward so slow

**Comment by johnma2006** [comment]
It's not meant to be fast! This repo is mostly only for educational purposes. I would suggest using the official repo to do any training: https://github.com/state-spaces/mamba

**Comment by D-Walter** [comment]
probably because mamba-ssm designed a GPU-adapted scanning operation based on C.

**Comment by A1phaCaO** [comment]
This model can only run on my CPU

**Comment by A1phaCaO** [comment]
This model can only run on my CPU
Python: 3.12.4
Pytorch: 2.3

**Comment by mengdeyu539** [comment]
I think the problem is embeding layer. D-layer is 50280, lol.
